bucket.go

```diff
@@ -109,75 +131,294 @@ type MutateInOptions struct {
 	GetAndTouchRaw(k string, exp uint32) (rv []byte, cas uint64, err error)
+
+	// Equivalent to GetAndTouchRaw, but does not return the value.
 	Touch(k string, exp uint32) (cas uint64, err error)
+
+	// Adds a document; similar to Set but gives up if the key exists with a non-nil value.
+	// Parameters:
+	// - k: The key (document ID)
+	// - exp: Expiration timestamp (0 for never)
+	// - v: The value to set. Will be marshaled to JSON unless it is a `[]byte` or `*[]byte`.
+	// Return values:
+	// - added: True if the document was added, false if it already has a value.
+	// - err: Error, if any. Does not return ErrKeyExists.
 	Add(k string, exp uint32, v interface{}) (added bool, err error)
+
+	// Adds a document; similar to SetRaw but gives up if the key exists with a non-nil value.
+	// Parameters:
+	// - k: The key (document ID)
+	// - exp: Expiration timestamp (0 for never)
+	// - v: The raw value to set.
+	// Return values:
+	// - added: True if the document was added, false if it already has a value.
+	// - err: Error, if any. Does not return ErrKeyExists.
 	AddRaw(k string, exp uint32, v []byte) (added bool, err error)
+
+	// Sets the value of a document, creating it if it doesn't exist.
+	// Parameters:
+	// - k: The key (document ID)
+	// - exp: Expiration timestamp (0 for never)
+	// - opts: Options. Use PreserveExpiry=true to leave the expiration alone
+	// - v: The value to set. Will be marshaled to JSON unless it is a `[]byte` or `*[]byte`
+	// Return values:
+	// - err: Error, if any
 	Set(k string, exp uint32, opts *UpsertOptions, v interface{}) error
+
+	// Sets the raw value of a document, creating it if it doesn't exist.
+	// Parameters:
+	// - k: The key (document ID)
+	// - exp: Expiration timestamp (0 for never)
+	// - opts: Options. Use PreserveExpiry=true to leave the expiration alone
+	// - v: The raw value to set
+	// Return values:
+	// - err: Error, if any. Does not return ErrKeyExists
 	SetRaw(k string, exp uint32, opts *UpsertOptions, v []byte) error
+
+	// The most general write method. Sets the value of a document, creating it if it doesn't
+	// exist, but checks for CAS conflicts:
+	// If the document has a value, and its CAS differs from the input `cas` parameter, the method
+	// fails and returns a CasMismatchErr.
+	// Parameters:
+	// - k: The key (document ID)
+	// - flags: TODO: What are they? Walrus ignores them
+	// - exp: Expiration timestamp (0 for never)
+	// - cas: Expected CAS value
+	// - v: The value to set. Will be marshaled to JSON unless it is a `[]byte` or `*[]byte`
+	// - opt: Options; see WriteOptions for details
+	// Return values:
+	// - casOut: The new CAS value
+	// - err: Error, if any. May be CasMismatchErr
 	WriteCas(k string, flags int, exp uint32, cas uint64, v interface{}, opt WriteOptions) (casOut uint64, err error)
+
+	// Deletes a document by setting its value to nil, making it a tombstone.
+	// System xattrs are preserved but user xattrs are removed.
```

Comment:
This feels like partial coverage of the xattr behaviour (doesn't describe that they are unchanged on WriteCas, deleted on Add, etc).  Could either add a note on each KVStore operation, or document elsewhere.

Reply:
That's good to know -- I wasn't aware of the behavior of Add, for example. I've added notes on both those methods. I'd like to add xattr notes on the other methods, but someone will need to tell me what the effects are.